File: V2/com.questvisionstream/src/rendering/DetectionMath.ts
import type { Detection, DetectionsPayload } from '../services/detection/types';
import type { NormalizedPoint, NormalizedRect, RenderBatch, RenderableDetection, Vec3 } from './types';

export interface NormalizeOptions {
  /**
   * Flip the Y axis. The Python server sends v-flipped frames (see
   * `VideoProcessor.flip_vertical`), so the Unity client used `invertY = true`.
   * Keep configurable to match the capture pipeline.
   */
  readonly invertY?: boolean;
  /**
   * Flip the X axis — for mirrored streams (front-facing cameras, or a server
   * running with `QVS_FLIP_HORIZONTAL=true`). Default false.
   */
  readonly invertX?: boolean;
}

/**
 * Convert a stream-pixel detection into a normalized viewport center + rect,
 * porting `DetectionSpawnerManager.ComputeWorldPosition`'s first stage:
 *
 *   nx = cx / frameW ; ny = cy / frameH ; if invertY: ny = 1 - ny
 *
 * The result is resolution-independent, so it stays correct as the server ramps
 * its frame size up over time (adaptive resolution).
 */
export function normalizeDetection(
  detection: Detection,
  frameWidth: number,
  frameHeight: number,
  options: NormalizeOptions = {},
): { center: NormalizedPoint; rect: NormalizedRect } {
  const invertY = options.invertY ?? true;
  const invertX = options.invertX ?? false;
  const w = Math.max(1, frameWidth);
  const h = Math.max(1, frameHeight);
  const [x1, y1, x2, y2] = detection.bbox;

  const cx = (x1 + x2) * 0.5;
  const cy = (y1 + y2) * 0.5;
  let nx = cx / w;
  let ny = cy / h;
  if (invertY) ny = 1 - ny;
  if (invertX) nx = 1 - nx;

  let rx = Math.min(x1, x2) / w;
  const rw = Math.abs(x2 - x1) / w;
  let ry = Math.min(y1, y2) / h;
  const rh = Math.abs(y2 - y1) / h;
  if (invertY) ry = 1 - ry - rh;
  if (invertX) rx = 1 - rx - rw;

  return { center: { x: nx, y: ny }, rect: { x: rx, y: ry, w: rw, h: rh } };
}

/** Build a full {@link RenderBatch} from a server payload. */
export function toRenderBatch(
  payload: DetectionsPayload,
  options: NormalizeOptions = {},
): RenderBatch {
  const detections: RenderableDetection[] = payload.detections.map((d) => {
    const { center, rect } = normalizeDetection(d, payload.width, payload.height, options);
    return { label: d.label, conf: d.conf, center, rect, source: d };
  });
  return {
    frame: payload.frame,
    frameWidth: payload.width,
    frameHeight: payload.height,
    detections,
  };
}

export type DedupPolicy = 'none' | 'per-class' | 'spatial-per-class';

export interface DedupOptions {
  readonly policy?: DedupPolicy;
  /** For 'spatial-per-class': min separation (metres) between same-class tags. */
  readonly minDistanceMeters?: number;
}

/**
 * Deduplication policy ported from the Unity client, which supported both
 * one-tag-per-class and spatial-per-class (`minDistanceMeters`) modes. Tracks
 * what has already been placed so repeat detections don't spawn duplicates.
 */
export class DetectionDeduper {
  private readonly policy: DedupPolicy;
  private readonly minDistance: number;
  private readonly placedClasses = new Set<string>();
  private readonly placedByClass = new Map<string, Vec3[]>();

  constructor(options: DedupOptions = {}) {
    this.policy = options.policy ?? 'per-class';
    this.minDistance = options.minDistanceMeters ?? 0.3;
  }

  /**
   * Should a detection of `label` at `worldPos` be placed? Records placement
   * when it returns true. `worldPos` is required only for 'spatial-per-class'.
   */
  shouldPlace(label: string, worldPos?: Vec3): boolean {
    switch (this.policy) {
      case 'none':
        return true;
      case 'per-class': {
        if (this.placedClasses.has(label)) return false;
        this.placedClasses.add(label);
        return true;
      }
      case 'spatial-per-class': {
        if (!worldPos) return true;
        const existing = this.placedByClass.get(label) ?? [];
        for (const p of existing) {
          if (distance(p, worldPos) < this.minDistance) return false;
        }
        existing.push(worldPos);
        this.placedByClass.set(label, existing);
        return true;
      }
    }
  }

  reset(): void {
    this.placedClasses.clear();
    this.placedByClass.clear();
  }
}

function distance(a: Vec3, b: Vec3): number {
  const dx = a.x - b.x;
  const dy = a.y - b.y;
  const dz = a.z - b.z;
  return Math.sqrt(dx * dx + dy * dy + dz * dz);
}

File: V2/com.questvisionstream.unity/Runtime/Core/DetectionMath.cs
// Copyright (c) Simon Jackson (SimonDarksideJ). All rights reserved.
// Licensed under the MIT License. See LICENSE in the repository root for license information.

using System.Collections.Generic;
using QuestVisionStream.Protocol;
using UnityEngine;

namespace QuestVisionStream.Core
{
    /// <summary>
    /// A rectangle in normalized viewport space (0..1, origin bottom-left).
    /// </summary>
    public readonly struct NormalizedRect
    {
        public NormalizedRect(float x, float y, float width, float height)
        {
            X = x;
            Y = y;
            Width = width;
            Height = height;
        }

        public float X { get; }
        public float Y { get; }
        public float Width { get; }
        public float Height { get; }

        public Vector2 BottomLeft => new Vector2(X, Y);
        public Vector2 BottomRight => new Vector2(X + Width, Y);
        public Vector2 TopLeft => new Vector2(X, Y + Height);
        public Vector2 TopRight => new Vector2(X + Width, Y + Height);
    }

    /// <summary>A detection converted to resolution-independent viewport space.</summary>
    public readonly struct RenderableDetection
    {
        public RenderableDetection(string label, float conf, Vector2 center, NormalizedRect rect, Detection source)
        {
            Label = label;
            Conf = conf;
            Center = center;
            Rect = rect;
            Source = source;
        }

        public string Label { get; }
        public float Conf { get; }

        /// <summary>Bbox center in normalized viewport space (0..1, origin bottom-left).</summary>
        public Vector2 Center { get; }

        public NormalizedRect Rect { get; }

        /// <summary>The raw stream-pixel detection this was derived from.</summary>
        public Detection Source { get; }
    }

    /// <summary>One payload's worth of renderable detections.</summary>
    public sealed class RenderBatch
    {
        public int Frame { get; internal set; }
        public long? Pts { get; internal set; }
        public int FrameWidth { get; internal set; }
        public int FrameHeight { get; internal set; }
        public IReadOnlyList<RenderableDetection> Detections { get; internal set; }
    }

    /// <summary>
    /// Stream-pixel to viewport-space conversion, ported from the V1
    /// <c>DetectionSpawnerManager.ComputeWorldPosition</c> first stage via the V2
    /// TypeScript <c>DetectionMath</c>:
    /// <code>nx = cx / frameW ; ny = cy / frameH ; if invertY: ny = 1 - ny</code>
    /// Normalizing against the PAYLOAD's width/height (never a constant) keeps the
    /// math correct as the server ramps its frame size (adaptive resolution).
    /// </summary>
    public static class DetectionMath
    {
        /// <summary>
        /// Convert a stream-pixel detection to a normalized viewport center + rect.
        /// </summary>
        /// <param name="detection">The raw detection.</param>
        /// <param name="frameWidth">Width of the frame the bbox was computed against.</param>
        /// <param name="frameHeight">Height of the frame the bbox was computed against.</param>
        /// <param name="invertY">
        /// Flip the Y axis. The server sends v-flipped frames by default
        /// (<c>QVS_FLIP_VERTICAL=true</c>), so this defaults on — keep it configurable
        /// to match the capture pipeline.
        /// </param>
        /// <param name="invertX">Flip the X axis, for mirrored streams. Default off.</param>
        public static void Normalize(
            in Detection detection,
            int frameWidth,
            int frameHeight,
            out Vector2 center,
            out NormalizedRect rect,
            bool invertY = true,
            bool invertX = false)
        {
            float w = Mathf.Max(1, frameWidth);
            float h = Mathf.Max(1, frameHeight);

            var cx = (detection.X1 + detection.X2) * 0.5f;
            var cy = (detection.Y1 + detection.Y2) * 0.5f;
            var nx = cx / w;
            var ny = cy / h;
            if (invertY) { ny = 1f - ny; }
            if (invertX) { nx = 1f - nx; }
            center = new Vector2(nx, ny);

            var rx = Mathf.Min(detection.X1, detection.X2) / w;
            var rw = Mathf.Abs(detection.X2 - detection.X1) / w;
            var ry = Mathf.Min(detection.Y1, detection.Y2) / h;
            var rh = Mathf.Abs(detection.Y2 - detection.Y1) / h;
            if (invertY) { ry = 1f - ry - rh; }
            if (invertX) { rx = 1f - rx - rw; }
            rect = new NormalizedRect(rx, ry, rw, rh);
        }

        /// <summary>Build a full <see cref="RenderBatch"/> from a server payload.</summary>
        public static RenderBatch ToRenderBatch(DetectionsPayload payload, bool invertY = true, bool invertX = false)
        {
            var detections = new List<RenderableDetection>(payload.Detections.Count);
            foreach (var detection in payload.Detections)
            {
                Normalize(detection, payload.Width, payload.Height, out var center, out var rect, invertY, invertX);
                detections.Add(new RenderableDetection(detection.Label, detection.Conf, center, rect, detection));
            }

            return new RenderBatch
            {
                Frame = payload.Frame,
                Pts = payload.Pts,
                FrameWidth = payload.Width,
                FrameHeight = payload.Height,
                Detections = detections
            };
        }
    }
}
